File: python/epoch/actor_id.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class ActorIdParts:
    region: int
    server: int
    process_type: int
    process_index: int
    actor_index: int


class ActorIdCodec(Protocol):
    def encode(self, parts: ActorIdParts) -> int:
        ...

    def decode(self, value: int) -> ActorIdParts:
        ...

    @property
    def name(self) -> str:
        ...


class DefaultActorIdCodec:
    _region_bits = 10
    _server_bits = 12
    _process_type_bits = 6
    _process_index_bits = 10
    _actor_index_bits = 26

    _region_max = (1 << _region_bits) - 1
    _server_max = (1 << _server_bits) - 1
    _process_type_max = (1 << _process_type_bits) - 1
    _process_index_max = (1 << _process_index_bits) - 1
    _actor_index_max = (1 << _actor_index_bits) - 1

    _actor_index_shift = 0
    _process_index_shift = _actor_index_shift + _actor_index_bits
    _process_type_shift = _process_index_shift + _process_index_bits
    _server_shift = _process_type_shift + _process_type_bits
    _region_shift = _server_shift + _server_bits

    @property
    def name(self) -> str:
        return "default"

    def encode(self, parts: ActorIdParts) -> int:
        if (parts.region < 0 or parts.region > self._region_max or
            parts.server < 0 or parts.server > self._server_max or
            parts.process_type < 0 or parts.process_type > self._process_type_max or
            parts.process_index < 0 or parts.process_index > self._process_index_max or
            parts.actor_index < 0 or parts.actor_index > self._actor_index_max):
            raise ValueError("ActorId parts out of range")

        return (
            (parts.region << self._region_shift)
            | (parts.server << self._server_shift)
            | (parts.process_type << self._process_type_shift)
            | (parts.process_index << self._process_index_shift)
            | (parts.actor_index << self._actor_index_shift)
        )

    def decode(self, value: int) -> ActorIdParts:
        return ActorIdParts(
            region=(value >> self._region_shift) & self._region_max,
            server=(value >> self._server_shift) & self._server_max,
            process_type=(value >> self._process_type_shift) & self._process_type_max,
            process_index=(value >> self._process_index_shift) & self._process_index_max,
            actor_index=(value >> self._actor_index_shift) & self._actor_index_max,
        )


default_actor_id_codec = DefaultActorIdCodec()


def encode_actor_id(parts: ActorIdParts, codec: ActorIdCodec = default_actor_id_codec) -> int:
    return codec.encode(parts)


def decode_actor_id(value: int, codec: ActorIdCodec = default_actor_id_codec) -> ActorIdParts:
    return codec.decode(value)

File: java/src/main/java/io/epoch/ActorId.java
package io.epoch;

public final class ActorId {
    private ActorId() {
    }

    public record Parts(int region, int server, int processType, int processIndex, int actorIndex) {
    }

    public interface Codec {
        long encode(Parts parts);

        Parts decode(long value);

        String name();
    }

    public static final Codec DEFAULT = new DefaultCodec();

    public static long encode(Parts parts) {
        return DEFAULT.encode(parts);
    }

    public static Parts decode(long value) {
        return DEFAULT.decode(value);
    }

    private static final class DefaultCodec implements Codec {
        private static final int REGION_BITS = 10;
        private static final int SERVER_BITS = 12;
        private static final int PROCESS_TYPE_BITS = 6;
        private static final int PROCESS_INDEX_BITS = 10;
        private static final int ACTOR_INDEX_BITS = 26;

        private static final long REGION_MAX = (1L << REGION_BITS) - 1;
        private static final long SERVER_MAX = (1L << SERVER_BITS) - 1;
        private static final long PROCESS_TYPE_MAX = (1L << PROCESS_TYPE_BITS) - 1;
        private static final long PROCESS_INDEX_MAX = (1L << PROCESS_INDEX_BITS) - 1;
        private static final long ACTOR_INDEX_MAX = (1L << ACTOR_INDEX_BITS) - 1;

        private static final int ACTOR_INDEX_SHIFT = 0;
        private static final int PROCESS_INDEX_SHIFT = ACTOR_INDEX_SHIFT + ACTOR_INDEX_BITS;
        private static final int PROCESS_TYPE_SHIFT = PROCESS_INDEX_SHIFT + PROCESS_INDEX_BITS;
        private static final int SERVER_SHIFT = PROCESS_TYPE_SHIFT + PROCESS_TYPE_BITS;
        private static final int REGION_SHIFT = SERVER_SHIFT + SERVER_BITS;

        @Override
        public long encode(Parts parts) {
            if (parts.region < 0 || parts.region > REGION_MAX ||
                parts.server < 0 || parts.server > SERVER_MAX ||
                parts.processType < 0 || parts.processType > PROCESS_TYPE_MAX ||
                parts.processIndex < 0 || parts.processIndex > PROCESS_INDEX_MAX ||
                parts.actorIndex < 0 || parts.actorIndex > ACTOR_INDEX_MAX) {
                throw new IllegalArgumentException("ActorId parts out of range");
            }
            return (long) parts.region << REGION_SHIFT
                | (long) parts.server << SERVER_SHIFT
                | (long) parts.processType << PROCESS_TYPE_SHIFT
                | (long) parts.processIndex << PROCESS_INDEX_SHIFT
                | (long) parts.actorIndex << ACTOR_INDEX_SHIFT;
        }

        @Override
        public Parts decode(long value) {
            int region = (int) ((value >> REGION_SHIFT) & REGION_MAX);
            int server = (int) ((value >> SERVER_SHIFT) & SERVER_MAX);
            int processType = (int) ((value >> PROCESS_TYPE_SHIFT) & PROCESS_TYPE_MAX);
            int processIndex = (int) ((value >> PROCESS_INDEX_SHIFT) & PROCESS_INDEX_MAX);
            int actorIndex = (int) ((value >> ACTOR_INDEX_SHIFT) & ACTOR_INDEX_MAX);
            return new Parts(region, server, processType, processIndex, actorIndex);
        }

        @Override
        public String name() {
            return "default";
        }
    }
}
